activemodel/lib/active_model/serialization.rb

```diff
@@ -123,7 +123,7 @@ module Serialization
     #   user.serializable_hash(include: { notes: { only: 'title' }})
     #   # => {"name" => "Napoleon", "notes" => [{"title"=>"Battle of Austerlitz"}]}
     def serializable_hash(options = nil)
-      attribute_names = self.attribute_names
+      attribute_names = respond_to?(:attribute_names) ? self.attribute_names : attributes.keys
```

Comment:
Instead of checking `respond_to` let's add a new method on active record for this.

`attributes_names_to_serialize`

Reply:
Ah, I guess the problem would be the same. I guess we need to find a better solution than the `respond_to` check which doesn't make this module override the Active Record implementation.

Reply:
Would this diff work?

```ruby
diff --git a/activemodel/lib/active_model/serialization.rb b/activemodel/lib/active_model/serialization.rb
index 81dbd0de16..b524f53817 100644
--- a/activemodel/lib/active_model/serialization.rb
+++ b/activemodel/lib/active_model/serialization.rb
@@ -123,7 +123,7 @@ module Serialization
     #   user.serializable_hash(include: { notes: { only: 'title' }})
     #   # => {"name" => "Napoleon", "notes" => [{"title"=>"Battle of Austerlitz"}]}
     def serializable_hash(options = nil)
-      attribute_names = respond_to?(:attribute_names) ? self.attribute_names : attributes.keys
+      attribute_names = attribute_names_for_serialization
 
       return serializable_attributes(attribute_names) if options.blank?
 
@@ -149,6 +149,9 @@ def serializable_hash(options = nil)
     end
 
     private
+      def attribute_names_for_serialization
+        attributes.keys
+      end
       # Hook method defining how an attribute value should be retrieved for
       # serialization. By default this is assumed to be an instance named after
       # the attribute. Override this method in subclasses should you need to
diff --git a/activerecord/lib/active_record/serialization.rb b/activerecord/lib/active_record/serialization.rb
index 212b6c4654..38f2ecfcf2 100644
--- a/activerecord/lib/active_record/serialization.rb
+++ b/activerecord/lib/active_record/serialization.rb
@@ -20,5 +20,11 @@ def serializable_hash(options = nil)
 
       super(options)
     end
+
+    private
+
+    def attribute_names_for_serialization
+      attribute_names
+    end
   end
 end
```

Reply:
Thanks @rafaelfranca - yes, that's a good approach. I've amended the code for that and re-done the benchmarks.